### PR TITLE
Malkov Websockets

### PR DESCRIPTION
**Pull request by dzinrai**

### CODE REVIEW

File: 18-websockets/classworks/Malkov_Aleksei/server.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+const cors = require('@koa/cors');
+app.use(cors());
+const server = require('http').createServer(app);
+const Router = require('koa-router');
+const router = new Router();
+const options = {origins:'*:*'};
+const io = require('socket.io')(server, {options});
+
+app.use(async (ctx, next) => {
+  await next();
```
Comment (Ivanko5417) [inline 18-websockets/classworks/Malkov_Aleksei/server.js]: image

File: 18-websockets/classworks/Malkov_Aleksei/server.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        username: socket.username,
+        numUsers: numUsers
+      });
+    }
+  });
+});
+
+router
+  .get('/', async (ctx, next) => {
+    console.log('hihi');
```
Comment (Ivanko5417) [inline 18-websockets/classworks/Malkov_Aleksei/server.js]: hehe

File: 18-websockets/classworks/Malkov_Aleksei/src/pages/Chat/Chat.jsx [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+  socket.on('new message', (data) => {
+    setListOfMessages([...listOfMessages, data]);
+  });
+
+  socket.on('typing', (data) => {
+    setTimeout(() => {
+      setTyping(null);
+    }, 1000)
+    setTyping(data);
```
Comment (Ivanko5417) [inline 18-websockets/classworks/Malkov_Aleksei/src/pages/Chat/Chat.jsx]: looks like it's impossible to see that a couple of users are typing

File: 18-websockets/classworks/Malkov_Aleksei/src/pages/Chat/Chat.jsx [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  });
+
+  function type(e) {
+    setMessage(e.target.value);
+    socket.emit("typing", username);
```
Comment (Ivanko5417) [inline 18-websockets/classworks/Malkov_Aleksei/src/pages/Chat/Chat.jsx:65]: if the user is a super typer man, you will get a huge number of requests from him. A good practice is to use the debounce method for calling function no more than one time per 2-4 secs.
Unfortunately, javascript doesn't have this method. You can write it by yourself or use some library.

**Review by Ivanko5417 — CHANGES_REQUESTED** [review]
Wonderful!

File: 18-websockets/classworks/Malkov_Aleksei/src/pages/Chat/Chat.jsx (hunk 4 above)
Comment (dzinrai) [reply 18-websockets/classworks/Malkov_Aleksei/src/pages/Chat/Chat.jsx:65]: i've added debounce method and multiple typing, but i think there is a problem with my timeouts
when im trying run 3 windows with react app, typing with one locking another 2 browser windows :D

Comment (Ivanko5417) [reply 18-websockets/classworks/Malkov_Aleksei/src/pages/Chat/Chat.jsx:65]: let's use library like that: https://www.npmjs.com/package/debounce

File: 18-websockets/classworks/Malkov_Aleksei/src/pages/Chat/Chat.jsx [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    }
+  });
+
+  function type(e) {
+    setMessage(e.target.value);
+    const emitter = debounce(function() {
+      socket.emit("typing", username);
+    }, 2000, true);
+    emitter();
+  }
```
Comment (Ivanko5417) [inline 18-websockets/classworks/Malkov_Aleksei/src/pages/Chat/Chat.jsx]: I think you should wrap the whole method `type` into `debounce`:
```javascript
const debouncedType = debounce(type, 2000, true)
```
Now your code will create a new function for every type and every function has it's own timers. I think that the problem

**Review by Ivanko5417 — CHANGES_REQUESTED** [review]

File: 18-websockets/classworks/Malkov_Aleksei/src/pages/Chat/Chat.jsx (hunk 5 above)
Comment (dzinrai) [reply 18-websockets/classworks/Malkov_Aleksei/src/pages/Chat/Chat.jsx]: socket.on() in component function body was a bad idea)) but its actually dont see updates of hooked states, so i put them in useEffects
with debounce npm and 
image
now browsers at least don't crash, ty)

File: 18-websockets/classworks/Malkov_Aleksei/src/pages/Chat/Chat.jsx [hunk 6]
```diff
@@ -23,34 +23,40 @@ function Chat() {
...
-    }
-  });
+  useEffect(() => {
+    socket.on('typing', (data) => {
+      //
+      if (!typing.includes(data.username)) {
+        setTimeout(() => {
+          setTyping([...typing.filter((typer) => typer !== data.username)]);
+        }, 2000)
+        setTyping([...typing, data.username]);
```
Comment (Ivanko5417) [inline 18-websockets/classworks/Malkov_Aleksei/src/pages/Chat/Chat.jsx:58]: Looks like there is one problem. You don't update user typing info if user has been already typing. For example, if user types and types again in one sec, other users will see typing message only for 2 seconds, not for 3.
The best way to fix this, it to reset the timer every time when you got an event about typing.

**Review by Ivanko5417 — CHANGES_REQUESTED** [review]
Almost done!

**Pull request merged**